### DISCUSSION TITLE
DTO $context instead of array $context

### DISCUSSION THREAD

**Discussion by Tachii** [Safemood/laravel-workflow#7] [Ideas]
Basically what the title says.

I think that it would be more usefull if `$context` would be a DTO so you could call `$context->param` instead of needing to check array keys like this: `issset($data['param') ?:`

- https://spatie.be/docs/laravel-data/v4/introduction

Pseudo code:

```php
use Spatie\LaravelData\Data;

class WorkflowContext extends Data implements DTO
{
    public function __construct(
        public string $param1,
        public int $param2,
        public ?string $optionalParam = null,
    ) {}
}

class SomeWorkflowStep
{
    public function execute(DTO $context)
    {
        // Instead of using isset($data['param']) or $data['param']
        // you can now directly access the properties like this:

        if ($context->param1) {
            // Do something with $context->param1
        }

        // Access other parameters as well
        if ($context->optionalParam) {
            // Do something with $context->optionalParam
        }
    }
}

// Usage example
$context = new WorkflowContext(
    param1: 'value1',
    param2: 123,
    optionalParam: 'optional'
);

$step = new SomeWorkflowStep();
$step->execute($context);

```

**Comment by Safemood** [comment]
Hi @Tachii,  

Thanks for the suggestion! I'm leaning towards keeping things lightweight, so I'll be avoiding additional dependencies like spatie/laravel-data. Instead, I'll opt for a custom, generic DTO implementation to maintain simplicity and flexibility without introducing extra dependencies.

**Comment by abbajbryant** [comment]
Would it be sufficient if the context type was widened out to include the Arrayable interface?

**Reply by Safemood** [reply]
Hi @abbajbryant ,

I'm not sure I fully understood your point. Could you share an example?

**Comment by abbajbryant** [comment]
I think it might be easiest to put together a small PR. But the general idea was to accept a context value that is array|Arrayable typed and handle turning it into the array representation as necessary in the Trait(s) that have function using the current array only context param.

**Reply by abbajbryant** [reply]
https://laravel.com/api/6.x/Illuminate/Contracts/Support/Arrayable.html

basically if it isn't an array call ->toArray on the Arrayble to get the object's array representation.

**Comment by Safemood** [comment]
Hey everyone,

I'm working on a PR to add a global DTOInterface for general use and a BaseDTO class. When BaseDTO is extended, it will automatically generate getters, setters, and implement a toArray method. This allows for minimalistic DTOs.

Even after this, you'll still have the freedom to either use the BaseDTO or implement your own approach with the DTOInterface.

```php
<?php
 
interface DTOInterface extends Arrayable
{
}

// Example : ProductDTO

namespace App\DTO;

class ProductDTO extends BaseDTO
{
    public bool $inStock;

    public function __construct(
        public string $name,
        public float $price,
        public int $quantity,
    ) {
        // Set the default stock status based on the quantity
        $this->inStock = $this->quantity > 0;
    }
}

// Usage Example
$productDTO = new ProductDTO(name: 'Laptop', price: 999.99, quantity: 5);


// Use the auto generated getters && setters methods
$productDTO->getQuantity(); // Output: 5

$productDTO->setQuantity(3);

$productDTO->setInStock(true);

$productDTO->getInStock()  // Output: bool(true)

// Convert the DTO to an array 
 $productDTO->toArray();
```
